### PR TITLE
Return to Mocha

### PR DESCRIPTION
**Pull request by danyalaytekin**
This PR reintroduces the [Mocha](https://mochajs.org/) test framework as a replacement for [Jest](https://jestjs.io/).  [Sinon](https://sinonjs.org/) and [Mockery](https://github.com/mfncooper/mockery) return too.  [Proclaim](https://github.com/rowanmanning/proclaim) returns temporarily pending a switch to a similar assertion library, since it has been sunsetted.  [c8](https://github.com/bcoe/c8) arrives as a drop-in replacement (maybe I should use quotes there but so far so good) for the dormant [nyc](https://www.npmjs.com/package/nyc).

The initial aim here was parity with both the Mocha implementation and the changes to the Jest implementation since.  The Mocha implementation was then upgraded with some Mocha improvements that have happened since then, such as [Mocha root hooks](https://mochajs.org/#root-hook-plugins) for parallelisation.

Credit to @joeyciechanowicz and @sangitamane for 2022's [PR going the other way](https://github.com/pa11y/pa11y/pull/622), because I'm sure those arguments and performance improvements absolutely made sense at the time, and the well-structured PR made this one much easier.

### Motivations

- Each project in the pa11y org now uses Mocha, which will aid with future plans.
- Overall, the Pa11y team's experience favours the use of Mocha.
- Mocha's parallelisation has removed the performance argument in favour of Jest.

#### While aiming for parity

- Reintroduce Mocha in place of Jest
    - Reversion of #622
- Reintroduce ~nyc~ c8 in place of Jest `--coverage`
    - using `c8` also fixes some [transformation warnings](https://github.com/istanbuljs/nyc/issues?q=is%3Aissue+is%3Aopen+transformation) which were affecting several tests (mentioned by #537) 
    - we weren't covering `bin/pa11y.js`, and I've left it outside coverage here
- Reintroduce Sinon and Mockery in place of Jest mocks (Mockery is stale and will be replaced in `pa11y@8.x`)
- Reintroduce Proclaim in place of Jest expectations

#### Further changes

- Use Mocha root hooks in place of setup/teardown files, to enable parallelisation
- Add [eslint-plugin-mocha](https://www.npmjs.com/package/eslint-plugin-mocha) and make code comply
- Extract Mocha configs into `.mocharc` files
- Extract some coverage config into `.c8rc` file
- Upgrade versions of these dependencies to latest
- Smuggled bump to `actions/setup-node` to `4` from `3`

### CODE REVIEW

**Review by sangitamane — APPROVED** [review]
Excellent!
Well done Dayal 🚀

**Review by hollsk — APPROVED** [review]
Thanks from me too, Danyal! 

Just the one question really, to satisfy my own curiousity

**Comment by danyalaytekin** [comment]
Thanks @sangitamane and @hollsk for the reviews 🙂 I'll have to rebase after merging the `AbortSignal` PR (or maybe forward merge... no, we should never do that... I really should've waited to merge this one first 😄) which might dismiss your reviews.  Will ping if so, and will also do another thorough scan before merging to make sure this hasn't lost any work that happened during the Jest era.

**Comment by danyalaytekin** [comment]
Hi @hollsk & @sangitamane, I've squashed the 33 commits and rebased the result on the [AbortSignal change](https://github.com/pa11y/pa11y/pull/694) that had arrived in `main`.  I've also

1. been more careful to retain the git histories of `test/unit/mocks/*` this time
1. removed `test/integration/global-setup.js`, because `test/integration/mocha-hooks.js` does its work now
1. looked over these changes again. I can see a few tighter assertions in a few places, such as object comparisons, but nothing significant and that might have been because it was the best way to compare them in Jest.  The Jest version was also reading and setting the integration test port explicitly, which doesn't happen in the Mocha version, but everything is working locally and in CI, so I think this is ok.

Here's a diff between this PR now and the last time you reviewed it:

- https://github.com/pa11y/pa11y/compare/return-to-mocha-presquash..return-to-mocha

File: test/unit/lib/action.test.js
Comment (hollsk) [reply test/unit/lib/action.test.js:10]: Oh my goodness. Yes, I can see why you chose this approach, thank you 😂 makes complete sense, and I think while the ideal world scenario is lovely and all it is a bit of a distraction at this stage, hehe

Thanks for explaining! Much appreciated ❤️

**Review by hollsk — APPROVED** [review]
Alright, let's go 🚀 🎸

**Pull request merged**
